wasm/func_validation.go

```diff
@@ -0,0 +1,948 @@
+			// then block.
+			valueTypeStack.resetAtStackLimit()
+		} else if op == OpcodeEnd {
+			bl := controlBloclStack[len(controlBloclStack)-1]
+			bl.endAt = pc
+			controlBloclStack = controlBloclStack[:len(controlBloclStack)-1]
+			if bl.isIf && bl.elseAt <= bl.startAt {
+				if len(bl.blockType.Results) > 0 {
+					return fmt.Errorf("type mismatch between then and else blocks")
+				}
+				// To handle if block without else properly,
+				// we set ElseAt to EndAt-1 so we can just skip else.
+				bl.elseAt = bl.endAt - 1
+			}
+			// Check type soundness.
+			if err := valueTypeStack.popResults(bl.blockType.Results, true); err != nil {
+				return fmt.Errorf("invalid instruction results at end instruction; expected %v: %v", bl.blockType.Results, err)
+			}
+			// Put the result types at the end after resetting at the stack limit
+			// since we might have Any type between the limit and the current top.
+			valueTypeStack.resetAtStackLimit()
+			for _, exp := range bl.blockType.Results {
+				valueTypeStack.push(exp)
+			}
+			// We exit if/loop/block, so reset the constraints on the stack manipulation
+			// on values previously pushed by outer blocks.
+			valueTypeStack.popStackLimit()
+		} else if op == OpcodeReturn {
+			expTypes := f.FunctionType.Type.Results
+			for i := 0; i < len(expTypes); i++ {
+				if err := valueTypeStack.popAndVerifyType(expTypes[len(expTypes)-1-i]); err != nil {
+					return fmt.Errorf("return type mismatch on return: %v; want %v", err, expTypes)
+				}
+			}
+			// return instruction is stack-polymorphic.
+			valueTypeStack.unreachable()
+		} else if op == OpcodeDrop {
+			_, err := valueTypeStack.pop()
+			if err != nil {
+				return fmt.Errorf("invalid drop: %v", err)
+			}
+		} else if op == OpcodeSelect {
+			if err := valueTypeStack.popAndVerifyType(ValueTypeI32); err != nil {
+				return fmt.Errorf("type mismatch on 3rd select operand: %v", err)
+			}
+			v1, err := valueTypeStack.pop()
+			if err != nil {
+				return fmt.Errorf("invalid select: %v", err)
+			}
+			v2, err := valueTypeStack.pop()
+			if err != nil {
+				return fmt.Errorf("invalid select: %v", err)
+			}
+			if v1 != v2 && v1 != valueTypeUnknown && v2 != valueTypeUnknown {
+				return fmt.Errorf("type mismatch on 1st and 2nd select operands")
+			}
+			if v1 == valueTypeUnknown {
+				valueTypeStack.push(v2)
+			} else {
+				valueTypeStack.push(v1)
+			}
+		} else if op == OpcodeUnreachable {
+			// unreachable instruction is stack-polymorphic.
+			valueTypeStack.unreachable()
+		} else if op == OpcodeNop {
+		} else {
+			return fmt.Errorf("invalid instruction 0x%x", op)
+		}
+	}
+
+	if len(controlBloclStack) > 0 {
+		return fmt.Errorf("ill-nested block exists")
+	}
+
+	if valueTypeStack.maximumStackPointer > maxStackValues {
+		return fmt.Errorf("function too large: potentially could have %d values on the stack with the limit %d",
+			valueTypeStack.maximumStackPointer, maxStackValues)
+	}
```

Comment:
this (and the addition of `maximumStackPointer` on `valueTypeStack`) is the only change in this commit